### ISSUE TITLE
negative loss

### ISSUE THREAD

**Issue by cizhouyu** [ShomyLiu/Pytorch-ENMF#1]
Hello! Thanks for contributing the pytorch version of ENMF.
When I followed the README file for model training, I found that the loss becomes negative, although the model indicator ndcg was still improving. Will similar negative loss appear in the [official tensorflow version](https://github.com/chenchongthu/ENMF) ? What are the possible reasons for the negative loss?
Hope to get your reply, thanks a lot!

![negative loss](https://github.com/ShomyLiu/Pytorch-ENMF/assets/60316519/084347d2-3c0d-470f-8033-a24d2c858cec)

**Comment by ShomyLiu** [comment]
Hi, Yes. The loss in ENMF could be negative, it is a relative value,  hence it will be ok if the loss becomes smaller.
And you could refer to the paper to find out the details of loss item.

**Comment by cizhouyu** [comment]
Thanks for your quick reply!

I have another question to consult: 
Regarding the loss calculation in each batch during training, in your code and the original paper of the author, each piece of data is the user (uid) and all the items (pos_iid) that user have interacted with in the training set. 
In the training process of many deep learning recommendation algorithms, each piece of data is user (uid) and item (iid). Can ENMF adopt such a training method? Or does it mean that ENMF must adopt the current training method?

Thanks! Looking forward to your reply

**Comment by cizhouyu** [comment]
> Hi, Yes. The loss in ENMF could be negative, it is a relative value, hence it will be ok if the loss becomes smaller. And you could refer to the paper to find out the details of loss item.

I see a constant term (const) in equation (11) in the paper; perhaps it is possible to make the loss always positive by adding a larger constant?  Although the loss function finally adopted is formula (14)

![image](https://github.com/ShomyLiu/Pytorch-ENMF/assets/60316519/cbee6336-5de3-4ba1-a277-93c20503ff84)

![image](https://github.com/ShomyLiu/Pytorch-ENMF/assets/60316519/3324ba92-7ee2-4a2b-aa0a-49d32fe97e3a)

**Comment by ShomyLiu** [comment]
Yes, the const value would result in a negative loss.  However, this does not affect the final model optimization.

**Comment by cizhouyu** [comment]
Maybe there is some ambiguity in my previous expression? I now understand that the model can be trained successfully even with a negative loss, but it seems more natural to add a constant so that the loss is always positive.

Do you have any suggestion about the feasibility of training ENMF model with another way of training (i.e. each piece of data is a user (uid) and an item (iid))? Thanks!
